lib/Utils/Container.cpp

```diff
@@ -97,46 +95,197 @@ MemRef<T, Dim>::MemRef(cv::Mat image, intptr_t offset, intptr_t sizes[Dim],
 }
 
-// Constructor for deep learning output.
-template <typename T, size_t Dim>
-MemRef<T, Dim>::MemRef(intptr_t results, intptr_t offset, intptr_t sizes[Dim],
-                       intptr_t strides[Dim]) {
-  auto ptr = new T[results];
-  this->allocated = ptr;
-  this->aligned = ptr;
-  for (int i = 0; i < results; i++) {
-    this->aligned[i] = 0;
+template <typename T, std::size_t N> MemRef<T, N>::~MemRef() {
+  if (allocated == aligned) {
+    if (allocated)
+      delete[] allocated;
+  } else {
+    if (allocated)
+      delete[] allocated;
+    if (aligned)
+      delete[] aligned;
   }
-  this->offset = offset;
-  for (int i = 0; i < Dim; i++)
-    this->sizes[i] = sizes[i];
-  for (int j = 0; j < Dim; j++)
-    this->strides[j] = strides[j];
 }
 
-template <typename T, size_t Dim> MemRef<T, Dim>::~MemRef() {
-  delete[] this->allocated;
+template <typename T, std::size_t N>
+void MemRef<T, N>::setStrides(const bool transpose) {
+  if (transpose) {
+    strides[0] = 1;
+    for (long i = 1; i < N; i++) {
+      strides[i] = strides[i - 1] * sizes[i - 1];
+    }
+  } else {
+    strides[N - 1] = 1;
+    for (long i = N - 2; i >= 0; i--) {
+      strides[i] = strides[i + 1] * sizes[i + 1];
+    }
+  }
+}
+
+template <typename T, std::size_t N>
+MemRef<T, N> MemRef<T, N>::transpose(const std::vector<size_t> &axes) {
+  static_assert(N >= 2 && N <= 4, "MemRef size not supported.");
+
+  std::vector<size_t> dims = axes;
+  if (dims.empty()) {
+    dims.resize(N);
+    std::iota(dims.rbegin(), dims.rend(), 0);
+  } else {
+    if (dims.size() != N) {
+      throw std::runtime_error("Invalid number of axes.");
+    }
+    std::vector<size_t> order = axes;
+    std::sort(order.begin(), order.end(),
+              [](size_t i, size_t j) { return i < j; });
+    if (order[0] != 0 || order[N - 1] != N - 1) {
+      throw std::runtime_error("All axes must be in range [0 ... N-1].");
+    }
+  }
+
+  intptr_t newSizes[N];
+  for (size_t i = 0; i < N; i++) {
+    newSizes[i] = sizes[dims[i]];
+  }
+  MemRef<T, N> res(newSizes);
+  const auto newStrides = res.getStrides();
+  // Copy the data
+  if (N == 2) {
+    for (std::size_t h = 0; h < sizes[0]; h++) {
+      for (std::size_t w = 0; w < sizes[1]; w++) {
+        size_t oldOffset = h * strides[0] + w * strides[1];
+        size_t newOffset = w * newStrides[0] + h * newStrides[1];
+        res[newOffset] = (*this)[oldOffset];
+      }
+    }
+  } else if (N == 3) {
+    auto getNewDim = [&dims](size_t index, size_t c, size_t h, size_t w) {
+      if (dims[index] == 0)
+        return c;
+      else if (dims[index] == 1)
+        return h;
+      else
+        return w;
+    };
+    for (std::size_t c = 0; c < sizes[0]; c++) {
+      for (std::size_t h = 0; h < sizes[1]; h++) {
+        for (std::size_t w = 0; w < sizes[2]; w++) {
+          size_t oldOffset = c * strides[0] + h * strides[1] + w * strides[2];
+          size_t newOffset = getNewDim(0, c, h, w) * newStrides[0] +
```

Comment:
newOffset is the offset of the transposed memref. It depends on the order of the dimensions of the transposed memref.
For example, let's say we want to transpose a chw memref to hwc ({0,1,2} -> {1, 2, 0}). The dimensions of the new memref are h, w and c instead of c, h, and w.

Reply:
In my understanding, you hope that the `getNewDim` function can help you set the dynamic c, h, w value, but you use a static value for the first parameter (the index). This means that the `getNewDim` function does not increase reusability, but is more redundant.

Reply:
Yes, that's right.
I have replaced the getNewDim function with an array.